app/state.go

```diff
@@ -731,6 +731,22 @@ func (uf *updateFetchState) Update() *datastore.UpdateInfo {
 	return &uf.update
 }
 
+func (uf *updateFetchState) HandleError(
+	ctx *StateContext,
+	c Controller,
+	err menderError,
+) (State, bool) {
+	log.Error(err.Error())
+
+	// Default for most update states. Some states will override this.
+	if uf.Update().SupportsRollback == datastore.RollbackSupported {
+		return NewUpdateRollbackState(uf.Update()), false
+	} else {
+		setBrokenArtifactFlag(ctx.Store, uf.Update().ArtifactName())
+		return NewUpdateErrorState(err, uf.Update()), false
+	}
```

Comment:
The `NewUpdateRollbackState` is not necessary, because as I mentioned earlier, when we are in the download state (or earlier), we have not changed the system yet, so no rollback is necessary. This means that `setBrokenArtifactFlag` is not necessary either, because we didn't change the system, so we know that the already installed artifact is still functioning. It's a rollback-by-doing-nothing, essentially. 🙂

So in fact I believe you can keep only the `NewUpdateErrorState` return, and that should be enough to solve the problem AFAICS!

I think this should also fix the unit test problem you are having.